File: 23-merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<pair<int,ListNode*>, vector<pair<int,ListNode*>>, greater<pair<int,ListNode*>>> pq;
        for(auto ele: lists){
            if(ele) pq.push({ele->val, ele});
        }
        ListNode *head = NULL;
        ListNode *prev = NULL;
        while(!pq.empty()){
            pair<int,ListNode*> p = pq.top();
            pq.pop();
            if(head == NULL){
                head = p.second;
                if(p.second->next) pq.push({p.second->next->val, p.second->next});
                prev = p.second;
            } else {
                prev->next = p.second;
                if(p.second->next) pq.push({p.second->next->val, p.second->next});
                prev = p.second;   
            }
        }

        return head;
    }
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:

        # Create dummy node
        dummyNode = ListNode(-1)
        
        # Create min heap
        pq = []
        heapq.heapify(pq)

        count = 0

        for node in lists:
            if node:
                heapq.heappush(pq, (node.val, count, node))
                count += 1
        
        curr = dummyNode
        while(len(pq) > 0):
            val, _, node = heapq.heappop(pq)
            curr.next = node
            if node.next:
                heapq.heappush(pq, (node.next.val, count, node.next))
                count += 1
            curr = curr.next
        
        return dummyNode.next

        